lightstep/sdk/metric/internal/syncstate/sync.go

```diff
@@ -0,0 +1,193 @@
+		// produce a viewstate.multiInstrment here.
+		compiled: viewstate.Combine(desc, compiled...),
+	}
+}
+
+// SnapshotAndProcess calls SnapshotAndProcess() for all live
+// accumulators of this instrument.  Inactive accumulators will be
+// subsequently removed from the map.
+func (inst *Instrument) SnapshotAndProcess() {
+	inst.current.Range(func(key interface{}, value interface{}) bool {
+		rec := value.(*record)
+		if rec.snapshotAndProcess() {
+			return true
+		}
+		// Having no updates since last collection, try to unmap:
+		if unmapped := rec.refMapped.tryUnmap(); !unmapped {
+			// The record is still referenced, continue.
+			return true
+		}
+
+		// If any other goroutines are now trying to re-insert this
+		// entry in the map, they are busy calling Gosched() awaiting
+		// this deletion:
+		inst.current.Delete(key)
+
+		// Last we'll see of this.
+		_ = rec.snapshotAndProcess()
+		return true
+	})
+}
+
+// record consists of an accumulator, a reference count, the number of
+// updates, and the number of collected updates.
+type record struct {
+	refMapped refcountMapped
+
+	// updateCount is incremented on every Update.
+	updateCount int64
+
+	// collectedCount is set to updateCount on collection,
+	// supports checking for no updates during a round.
+	collectedCount int64
+
+	// accumulator can be a multi-accumulator if there
+	// are multiple behaviors or multiple readers, but
+	// these distinctions are not relevant for synchronous
+	// instruments.
+	accumulator viewstate.Accumulator
+}
+
+// snapshotAndProcessRecord checks whether the accumulator has been
+// modified since the last collection (by any reader), returns a
+// boolean indicating whether the record is active.  If active, calls
+// SnapshotAndProcess on the associated accumulator and returns true.
+// If updates happened since the last collection (by any reader),
+// returns false.
+func (rec *record) snapshotAndProcess() bool {
+	mods := atomic.LoadInt64(&rec.updateCount)
+	coll := atomic.LoadInt64(&rec.collectedCount)
+
+	if mods == coll {
+		return false
+	}
+	// Updates happened in this interval, collect and continue.
+	atomic.StoreInt64(&rec.collectedCount, mods)
+
+	rec.accumulator.SnapshotAndProcess()
+	return true
+}
+
+// capture performs a single update for any synchronous instrument.
+func capture[N number.Any, Traits number.Traits[N]](_ context.Context, inst *Instrument, num N, attrs []attribute.KeyValue) {
+	if inst.compiled == nil {
+		return
+	}
+
+	// Note: Here, this is the place to use context, e.g., extract baggage.
+
+	rec, updater := acquireRecord[N](inst, attrs)
+	defer rec.refMapped.unref()
+
+	updater.Update(num)
+
+	// Record was modified.
+	atomic.AddInt64(&rec.updateCount, 1)
+}
+
+// acquireRecord gets or creates a `*record` corresponding to `attrs`,
+// the input attributes.
+func acquireRecord[N number.Any](inst *Instrument, attrs []attribute.KeyValue) (*record, viewstate.Updater[N]) {
+	aset := attribute.NewSet(attrs...)
```

Comment:
Same here, is there a good way for us to prep if we want to use fingerprinting? More likely than not I'll PR something up after the fact.

Reply:
I did prototype fingerprinting once. The result is heavy changes in this file. The way Prometheus golang's client does this, the `sync.Map` would be replaced by a `map[fingerprint][]*record` and a `sync.RWMutex`. The refcount_mapped stuff can be deleted, then.

Another way keeps the refcount_mapped stuff, leaves the map as `map[fingerprint]*record` but allows for an embedded linked list of duplicates, i.e., each record has a `next unsafe.Pointer` field (actually an atomic `*record`) for duplicates. This avoids a slice in the common case, also makes it simple to have a "ignore duplicates / permit collisions" setting that avoids checking for exact matches after fingerprint matches. The "remove from map" logic would only remove from a map when all collisions in a bucket are unused, which leaves room for pathological performance cases. Let's discuss / study this in another PR.